services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java

```diff
@@ -292,6 +225,44 @@ protected Response exchangeStoredToken(UriInfo uriInfo, EventBuilder event, Clie
         }
     }
 
+    private String getIDTokenForLogout(KeycloakSession session, UserSessionModel userSession) {
+        String tokenExpirationString = userSession.getNote(FEDERATED_TOKEN_EXPIRATION);
+        long exp = tokenExpirationString == null ? 0 : Long.parseLong(tokenExpirationString);
+        int currentTime = Time.currentTime();
+        if (exp > 0 && currentTime > exp) {
+            String response = refreshTokenForLogout(session, userSession);
+            AccessTokenResponse tokenResponse = null;
+            try {
+                tokenResponse = JsonSerialization.readValue(response, AccessTokenResponse.class);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            return tokenResponse.getIdToken();
+        } else {
+            return userSession.getNote(FEDERATED_ID_TOKEN);
+
+        }
+    }
+
+    protected void processAccessTokenResponse(BrokeredIdentityContext context, AccessTokenResponse response) {
+
+
+    }
+
+    protected SimpleHttp getRefreshTokenRequest(KeycloakSession session, String refreshToken, String clientId, String clientSecret) {
+        if (OIDCLoginProtocol.CLIENT_SECRET_BASIC.equals(getConfig().getClientAuthMethod())) {
```

Comment:
Bonus point: If you're able to add stuff from AbstractOAuth2IdentityProvider.generateTokenRequest to support JWT authentication here, it will be great! Looks like this capability is missing in the PR, which added the support for JWT authentication. Thanks